src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs

```diff
@@ -231,15 +289,16 @@ protected virtual void OnConsumedBytes(int count)
             // see also http://tools.ietf.org/html/rfc2616#section-4.4
             var keepAlive = httpVersion != HttpVersion.Http10;
 
-            var connection = headers.HeaderConnection.ToString();
-            if (connection.Length > 0)
+            var connection = headers.HeaderConnection;
+            if (connection.Count > 0)
             {
-                if (connection.Equals("upgrade", StringComparison.OrdinalIgnoreCase))
+                bool upgrade;
+                ParseConnection(connection, out keepAlive, out upgrade);
+
+                if (upgrade)
                 {
-                    return new ForRemainingData(context);
+                    return new ForRemainingData(upgrade, context);
```

Comment:
Did you decide what to do yet about making sure the Request.Body is not readable on Upgrade requests?